fence_token_storage.py

```diff
@@ -0,0 +1,157 @@
+         much time locking.
+        :param fence_fetch_fn: The function to fetch the credentials from the fence. This function will ensure that
+        fence_fetch_fn is not called multiple times concurrently. Arguments should work as
+        fence_fetch_fn(prep_key_fn(key)) -> returns the string fence account credentials.
+        :return returns the fence service account json key and the expiration time for that value.
+        """
+        fence_service_account = fsa_key.get()
+        now = datetime.datetime.now()
+        if fence_service_account is None or \
+                fence_service_account.expires_at is None or \
+                fence_service_account.expires_at < now:
+            fence_service_account = self._fetch_service_account(fsa_key, prep_key_fn, fence_fetch_fn)
+
+        return (fence_service_account.key_json, fence_service_account.expires_at)
+
+    def _fetch_service_account(self, fsa_key, prep_key_fn, fence_fetch_fn):
+        """
+        Fetch a new service account from fence. We must be careful that concurrent requests result in only one
+        key request to fence so the service account does not run out of keys (google limits to 10).
+        """
+        # Prep key before acquiring lock to keep lock duration as small as possible.
+        prepped_key = prep_key_fn(fsa_key)
+
+        if self._acquire_lock(fsa_key):
+            key_json = fence_fetch_fn(prepped_key)
+            fence_service_account = FenceServiceAccount(key_json=key_json,
+                                                        expires_at=datetime.datetime.now() + _FSA_KEY_LIFETIME,
+                                                        update_lock_timeout=None,
+                                                        key=fsa_key)
+            fence_service_account.put()
+
+        else:
+            fence_service_account = self._wait_for_update(fsa_key)
+            if not fence_service_account.key_json or (
+                    fence_service_account.expires_at and fence_service_account.expires_at < datetime.datetime.now()):
+                # We waited for a fence service account update since someone else was holding the lock, but the
+                # lock expired without a valid update.
+                # we could recursively call _fetch_service_account_json at this point but let's start with failure
+                raise ServiceAccountNotUpdatedException(
+                    "lock on key {} expired but value was not updated".format(fsa_key))
+
+        return fence_service_account
+
+    def _acquire_lock(self, fsa_key):
+        """
+        :param fsa_key:
+        :return: True if the lock was acquired, False otherwise
+        """
+        try:
+            return self._lock_fence_service_account(fsa_key)
+        # We expect a transaction failure when someone else acquires the lock instead of us. That's fine, it's just
+        # a different way we could fail to acquire the lock.
+        except TransactionFailedError:
```

Comment:
I scoped down this exception to a transaction failure, removing the old TODO

Reply:
did you verify this is correct? I think part of the problem was that is was unclear which are the exceptions that should be handled

Reply:
That was my reading of the docs here[0] but I did not try to provoke a transaction error. Looking some more, it sounds pretty ambiguous in the note[1], so I'll go back to `except Exception`

[0] https://cloud.google.com/appengine/docs/standard/python/datastore/transactions#what_can_be_done_in_a_transaction
[1] https://cloud.google.com/appengine/docs/standard/python/datastore/transactions#using_transactions